cooked-validators/tests/Cooked/AttackSpec.hs

```diff
@@ -0,0 +1,330 @@
+      testCase "careful minting policy" $
+        let tName = L.tokenName "MockToken"
+            pol = carefulPolicy tName 1
+         in testFailsFrom'
+              isCekEvaluationFailure
+              def
+              ( somewhere
+                  (dupTokenAttack (\_ n -> Just $ n + 1) (wallet 6))
+                  (dupTokenTrace pol tName 1 (wallet 1))
+              ),
+      testCase "careless minting policy" $
+        let tName = L.tokenName "MockToken"
+            pol = carelessPolicy
+         in testSucceeds $
+              somewhere
+                (dupTokenAttack (\_ n -> Just $ n + 1) (wallet 6))
+                (dupTokenTrace pol tName 1 (wallet 1))
+    ]
+
+-- * Tests for the datum hijacking attack
+
+-- ** Mock contract for the datum hijacking attack
+
+-- This is a very simple contract: The first transaction locks some Ada to the
+-- validator, using the datum 'FirstLock', the second transaction then re-locks
+-- the same amount to the same validator, using the datum 'SecondLock'. The
+-- datum hijacking attack should target the secon transaction, and substitute a
+-- different recipient.
+
+data MockDatum = FirstLock | SecondLock deriving (Show)
+
+instance Pl.Eq MockDatum where
+  {-# INLINEABLE (==) #-}
+  FirstLock == FirstLock = True
+  SecondLock == SecondLock = True
+  _ == _ = False
+
+Pl.makeLift ''MockDatum
+Pl.unstableMakeIsData ''MockDatum
+
+data MockContract
+
+instance L.ValidatorTypes MockContract where
+  type DatumType MockContract = MockDatum
+  type RedeemerType MockContract = ()
+
+-- ** Transactions (and 'TxSkels') for the datum hijacking attack
+
+lockValue :: L.Value
+lockValue = L.lovelaceValueOf 12345678
+
+lockTxSkel :: SpendableOut -> L.TypedValidator MockContract -> TxSkel
+lockTxSkel o v =
+  txSkelOpts
+    (def {adjustUnbalTx = True})
+    ([SpendsPK o] :=>: [PaysScript v FirstLock lockValue])
+
+txLock :: MonadBlockChain m => L.TypedValidator MockContract -> m ()
+txLock v = do
+  me <- ownPaymentPubKeyHash
+  utxo : _ <- pkUtxosSuchThatValue me (`L.geq` lockValue)
+  void $ validateTxSkel $ lockTxSkel utxo v
+
+relockTxSkel :: L.TypedValidator MockContract -> SpendableOut -> TxSkel
+relockTxSkel v o =
+  txSkelOpts
+    (def {adjustUnbalTx = True})
+    ( [SpendsScript v () (o, FirstLock)]
+        :=>: [PaysScript v SecondLock lockValue]
+    )
+
+txRelock ::
+  MonadBlockChain m =>
+  L.TypedValidator MockContract ->
+  m ()
+txRelock v = do
+  utxo : _ <- scriptUtxosSuchThat v (\d _ -> FirstLock Pl.== d)
+  void $ validateTxSkel $ relockTxSkel v (fst utxo)
+
+-- ** Validators for the datum hijacking attack
+
+-- | Try to extract a datum from an output.
+{-# INLINEABLE outputDatum #-}
+outputDatum :: L.TxInfo -> L.TxOut -> Maybe MockDatum
+outputDatum txi o = do
+  h <- L.txOutDatum o
+  L.Datum d <- L.findDatum h txi
+  Pl.fromBuiltinData d
+
+{-# INLINEABLE mkCarefulValidator #-}
+mkCarefulValidator :: MockDatum -> () -> L.ScriptContext -> Bool
```

Comment:
It seems lie `mkCarefulValidator` and `mkCarelessValidator` could be abstracted through a common `mkMockValidator`

```haskell
mkCarefulValidator = mkMockValidator L.getContinuingOutputs
mkCarelessValidator = mkMockValidator (L.txInfoOutputs . L.scriptContextTxInfo)
```